doc/architecture-decisions/0012-routecontroller-route-deletion-finalizer.md

```diff
@@ -0,0 +1,27 @@
+# 11. Routecontroller Route Deletion Finalizer
+
+Date: 2020-05-11
+
+## Status
+
+Accepted
+
+## Context
+Without a finalizer, we could rely on cascading deletion of child objects when Routes are deleted and `RequeueAfter` to rebuild the virtual services and services  from its routes every `ResyncInterval` seconds. `ResyncInterval` is currently set to 30 seconds.
+
+However, this doesn’t meet our SLO to handle changes within 10 seconds. Cascading deletes don’t work within 10 seconds, so meeting the SLO would require `ResyncInterval` to be less than 10 seconds, which seems unreasonable. 
+
+Cascading deletion alone also doesn’t handle the case of a virtual service being owned by >1 Route. This means a cascading delete cannot update the virtual service’s contents to not include the paths related to that deleted route. So we would need to rely on `RequeueAfter` for those updates, which would be slow.
+
+
+## Decision
+
+In order to handle all of the cases: deleting services, deleting virtual services owned by only that route and updating virtual services owned by many routes, we rely on a finalizer, so we can have a “fast path” to all of these cases. 
+
+Finalizers do a “soft delete” to keep the route in the K8s API while handling deletion/updates to the route’s child objects.  
+
+Using finalizers as a “fast path” and `RequeueAfter` as a “sync” to handle any failures can help us meet our SLO of having 95% of route changes being reflected within 10 seconds.
```

Comment:
I think it is a bit dangerous to think of `RequeueAfter` as being a part of the "slow-path" of our reconciliation logic and I would prefer we phrase it as a "sync" for disaster recovery purposes. Like it's there in case something unexpected happens outside of our normal operation (e.g. `etcd` losing child resources) so that the system can heal without the `Route`s needing to be updated. But we shouldn't rely on it for any of our business logic needs. In the case of the `VirtualService` update happening in the finalizer I would consider it as the _only_ path.

🤔 🤔 

This thought probably needs its own ADR or something, to be honest. @christianang mentioned something about we could rely on Kubebuilder's `Owns` functionality to also watch child resources, but that requires us being able to set the `OwnerReference.Controller` attribute to `true` which caused us problems... something we're going to poke at and explore a bit more I think... this is why we haven't written _that_ ADR. 😂 

Sorry for the ramble.

Reply:
for just this ADR, I think I can do some rephrasing to make it clear that finalizers help implement all of the cases in our route deletion reconciliation logic (and not call it a fast path), and add that `RequeueAfter` is for disaster recovery rather than "any failures" as written above. 

Is the `Owns` functionality more along the lines of a "slow path" that we should explore? Would it help tighten the gap between our business logic (all of our reconcilation logic including finalizers) and disaster recovery?

Reply:
Owns is kind of like a fast path for recovery. It would make the controller watch for changes to child resources of a `Route` and trigger a reconcile if anything happens to those. So for like our example where someone accidentally `kubectl delete`s all the `Services` the controller would re-reconcile the corresponding Routes.

They touch a bit on it here in the `CronJob` example:
https://book.kubebuilder.io/cronjob-tutorial/controller-implementation.html#setup

It's unclear to us how well it works for the `VirtualService` case and if it fully covers all of the things `RequeueAfter` protects us from.